src/encoder.rs

```diff
@@ -61,64 +66,97 @@ where
     }
 
     pub fn with_chunks_size(output: W, chunks: usize) -> Encoder<W> {
-        Encoder {
+        let chunks_size = if chunks > MAX_CHUNK_SIZE {
+            MAX_CHUNK_SIZE
+        } else {
+            chunks
+        };
+        let mut encoder = Encoder {
             output,
-            chunks_size: chunks,
+            chunks_size,
             buffer: Vec::with_capacity(0),
             flush_after_write: false,
-        }
+        };
+        encoder.reset_buffer();
+        encoder
     }
 
     pub fn with_flush_after_write(output: W) -> Encoder<W> {
-        Encoder {
+        let mut encoder = Encoder {
             output,
             chunks_size: 8192,
             buffer: Vec::with_capacity(0),
             flush_after_write: true,
-        }
+        };
+        encoder.reset_buffer();
+        encoder
     }
-}
 
-fn send<W>(output: &mut W, data: &[u8]) -> IoResult<()>
-where
-    W: Write,
-{
-    write!(output, "{:x}\r\n", data.len())?;
-    output.write_all(data)?;
-    write!(output, "\r\n")?;
-    Ok(())
+    fn reset_buffer(&mut self) {
+        // Reset buffer
+        self.buffer.clear();
+        // Arbitrary bytes to take up space. Should never be seen in output.
+        let placeholder: [u8; HEADER_SIZE] = *b"DACE\r\n";
+        self.buffer.extend_from_slice(&placeholder);
```

Comment:
are these the lines responsible for the optimization? how would this differ from doing `self.buffer.reserve(HEADER_SIZE)`?

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reserve

then we wouldn't need a `buffer_empty` method since we can just do `self.buffer.is_empty()`

Reply:
These lines, along with the code in `send` that writes the prelude into the reserved space. Here's how I think of it: In order to send the whole chunk in one `write` call, we need a slice of contiguous bytes that contain the chunk size, `\r\n`, the chunk body, then `\r\n` again. One way to get that is to write into the buffer as normal, then when we're ready to write, moving all those bytes rightward to make room for the chunk size.

However, that's a little bit inefficient - it means potentially moving a lot of bytes around. Instead, we reserve space for the chunk size from the very beginning, so we don't have to move anything when it's time to call write. I think `reserve` doesn't work for this purpose, since it makes space available at the end rather than at the beginning.

Reply:
By the way, I had another thought to perhaps express the intent more clearly in code: I could set the initial contents to `format!("{:x}\r\n", MAX_CHUNK_SIZE)`, which is also six bytes.

Reply:
thanks for the write-up. my brain is a little fried right now, but i'll revisit this again tomorrow and let you know if i have any follow-up q's